### PR TITLE
[WIP] macOS support

### PR DESCRIPTION
**Pull request by michaelsbradleyjr**
This is my attempt to catch up @grendello's fork's `macos` branch with this repo's current `master`, cf. PR #1280.

Note: I'm not a C programmer (will learn as I go) and only in the last year+ have been getting familiar with "systems" programming, Makefiles, cmake, etc., after spending many years working with langs/envs such as Node.js, Clojure, etc. In other words, my changes in this draft PR don't represent a deep analysis or contribution, I mainly just fiddled with @grendello's modifications until I got them working again.

With these changes in place and a fresh clone (and on this PR branch), the following works on macOS (I'm using 10.15), as long as the necessaries have been installed with `brew`:
```
$ mkdir build && cd build
$ cmake -DCMAKE_BUILD_TYPE=Release -DUSE_MULTIMEDIA=ffmpeg ..
$ make
```

After building, I've tried running `build/notcurses-demo` in iTerm2 (`TERM=xterm-256color`), Alacritty (`TERM=alacritty`), and Kitty (`TERM=xterm-kitty`) on macOS (`COLORTERM=truecolor`):
```
$ build/notcurses-demo -p ./data
```

In all cases the demo doesn't run all the way through, always ends like this:
```
...
14│ animate│ 28.43ms│     11│ 123.13Ki│  386.9│ 7│ 1│71│ 482.62║FAILED
15│    reel│   0.00s│      0│     0.00│    0.0│ 0│ 0│ 0│   0.00║SKIPPED
16│whiteout│   0.00s│      0│     0.00│    0.0│ 0│ 0│ 0│   0.00║SKIPPED
17│uniblock│   0.00s│      0│     0.00│    0.0│ 0│ 0│ 0│   0.00║SKIPPED
18│    view│   0.00s│      0│     0.00│    0.0│ 0│ 0│ 0│   0.00║SKIPPED
19│   luigi│   0.00s│      0│     0.00│    0.0│ 0│ 0│ 0│   0.00║SKIPPED
20│ sliders│   0.00s│      0│     0.00│    0.0│ 0│ 0│ 0│   0.00║SKIPPED
21│ fallin'│   0.00s│      0│     0.00│    0.0│ 0│ 0│ 0│   0.00║SKIPPED
22│  jungle│   0.00s│      0│     0.00│    0.0│ 0│ 0│ 0│   0.00║SKIPPED
23│  qrcode│   0.00s│      0│     0.00│    0.0│ 0│ 0│ 0│   0.00║SKIPPED
24│     zoo│   0.00s│      0│     0.00│    0.0│ 0│ 0│ 0│   0.00║SKIPPED
25│   outro│   0.00s│      0│     0.00│    0.0│ 0│ 0│ 0│   0.00║SKIPPED
══╧════════╧════════╪═══════╪═════════╪═══════╧══╧══╧══╧═══════╝
...
Error running demo.
Is "./data" the correct data path? Supply it with -p.
```

Always reported as not available: `sexblitter`.
Reported as not available in iTerm2 and Alacritty: `quadblitter`, `sexblitter`, `pixelblitter`.

I'm not sure why it doesn't think `./data` is the correct data path for some parts of the demo. Note that I didn't install with `make install`, preferring to run from build and data in the repo.

The framerate is low-ish in iTerm2 and better in Alacritty, the visual quality is about the same in either. In Kitty the visual quality is much better but the framerate is abysmal (about 1 to 2 fps) during the intro with the spinning 3D text and also when the world map is being colored red.

I've tried compiling with the default compiler, i.e. clang that ships with Xcode. I've also tried with compilers installed via brew:
```
export CC=/usr/local/opt/gcc/bin/gcc-11
export CXX=/usr/local/opt/gcc/bin/g++-11
```
```
export CC=/usr/local/opt/llvm/bin/clang-12
export CXX=/usr/local/opt/llvm/bin/clang-12
```
I didn't notice any differences from the results described above.

Here's the tail end of the results of running `./notcurses-tester -p ../data` in Kitty:
```
                                                             ncdirect_core_init:737:Passed unsupported flags 0xffffffffffffffff
Signals weren't registered for 0x7f860a00c000 (had 0x0)
===============================================================================
/Users/michael/repos/forks/notcurses/src/tests/egcpool.cpp:5:
TEST CASE:  EGCpool
  AddAndRemove

/Users/michael/repos/forks/notcurses/src/tests/egcpool.cpp:51: ERROR: CHECK( 1 == nccell_width(n_, &c) ) is NOT correct!
  values: CHECK( 1 == 0 )

/Users/michael/repos/forks/notcurses/src/tests/egcpool.cpp:52: FATAL ERROR: REQUIRE( 0 <= egcpool_stash(&pool_, wstr, ulen) ) is NOT correct!
  values: REQUIRE( 0 <= -1 )

get_tty_fd:490:Returning TTY fd 4
setup_signals:114:0x7f860a011600 is already registered for signals (provided 0x7f860a024000)
===============================================================================
/Users/michael/repos/forks/notcurses/src/tests/Exceptions.cpp:7:
TEST CASE:  Exceptions
  ResetStats

/Users/michael/repos/forks/notcurses/src/tests/Exceptions.cpp:7: ERROR: test case THREW exception: exception thrown in subcase - will translate later when the whole test case has been exited (cannot translate while there is an active exception)

===============================================================================
/Users/michael/repos/forks/notcurses/src/tests/Exceptions.cpp:7:
TEST CASE:  Exceptions

DEEPEST SUBCASE STACK REACHED (DIFFERENT FROM THE CURRENT ONE):
  ResetStats

/Users/michael/repos/forks/notcurses/src/tests/Exceptions.cpp:7: ERROR: test case THREW exception: unknown exception

get_tty_fd:490:Returning TTY fd 13
setup_signals:114:0x7f860a011600 is already registered for signals (provided 0x7f860a024000)
===============================================================================
/Users/michael/repos/forks/notcurses/src/tests/Ncpp.cpp:7:
DESCRIPTION: Basic C++ wrapper tests
TEST CASE:  Ncpp
  ConstructNotCurses

/Users/michael/repos/forks/notcurses/src/tests/Ncpp.cpp:7: ERROR: test case THREW exception: exception thrown in subcase - will translate later when the whole test case has been exited (cannot translate while there is an active exception)

===============================================================================
/Users/michael/repos/forks/notcurses/src/tests/Ncpp.cpp:7:
DESCRIPTION: Basic C++ wrapper tests
TEST CASE:  Ncpp

DEEPEST SUBCASE STACK REACHED (DIFFERENT FROM THE CURRENT ONE):
  ConstructNotCurses

/Users/michael/repos/forks/notcurses/src/tests/Ncpp.cpp:7: ERROR: test case THREW exception: unknown exception

===============================================================================
/Users/michael/repos/forks/notcurses/src/tests/sixel.cpp:140:
TEST CASE:  Sixels

/Users/michael/repos/forks/notcurses/src/tests/sixel.cpp:142: FATAL ERROR: REQUIRE( nullptr != nc_ ) is NOT correct!
  values: REQUIRE( NULL != NULL )

===============================================================================
/Users/michael/repos/forks/notcurses/src/tests/version.cpp:3:
TEST CASE:  Version

/Users/michael/repos/forks/notcurses/src/tests/version.cpp:5: FATAL ERROR: REQUIRE( nullptr != nc_ ) is NOT correct!
  values: REQUIRE( NULL != NULL )

===============================================================================
/Users/michael/repos/forks/notcurses/src/tests/visual.cpp:5:
TEST CASE:  Visual

/Users/michael/repos/forks/notcurses/src/tests/visual.cpp:7: FATAL ERROR: REQUIRE( nullptr != nc_ ) is NOT correct!
  values: REQUIRE( NULL != NULL )

===============================================================================
/Users/michael/repos/forks/notcurses/src/tests/wide.cpp:10:
TEST CASE:  Wide

/Users/michael/repos/forks/notcurses/src/tests/wide.cpp:12: FATAL ERROR: REQUIRE( nullptr != nc_ ) is NOT correct!
  values: REQUIRE( NULL != NULL )

===============================================================================
/Users/michael/repos/forks/notcurses/src/tests/zaxis.cpp:5:
TEST CASE:  ZAxis

/Users/michael/repos/forks/notcurses/src/tests/zaxis.cpp:7: FATAL ERROR: REQUIRE( nullptr != nc_ ) is NOT correct!
  values: REQUIRE( NULL != NULL )

===============================================================================
[doctest] test cases:    47 |    38 passed |  9 failed | 1 skipped
[doctest] assertions: 13658 | 13646 passed | 12 failed |
[doctest] Status: FAILURE!
```

---

Notcurses seems like a great project, truly inspiring! In the near-ish future I hope to contribute a [Nim](https://nim-lang.org/) language wrapper, as it happens that my current employment has me learning and programming in Nim on a daily basis. I'd like to write a killer TUI app as an alternative/complement to [status-desktop](https://github.com/status-im/status-desktop) (I also work on that project) and at present am experimenting with a Nim wrapper around ncurses. I think it would be a lot more interesting and exciting to implement that TUI app with notcurses! 🚀 

In addition to working with notcurses on macOS, I'll see how far I can get compiling it in an MSYS2 context and running the demo in Alacritty, etc. on MS Windows.

### CODE REVIEW

**Comment by dankamongmen** [comment]
exciting shit! i have a mac now, though it might be a bit older than is desirable. tell me when you think i ought start looking at this, and i can try to build it. it'll likely take me some time to get there, as i've never developed on a mac.

and thank you for your kind words!

**Comment by michaelsbradleyjr** [comment]
> exciting shit! i have a mac now, though it might be a bit older than is desirable. tell me when you think i ought start looking at this, and i can try to build it. it'll likely take me some time to get there, as i've never developed on a mac.

Is it the mac mentioned in https://github.com/dankamongmen/notcurses/pull/1280#issuecomment-780911308?

**Comment by dankamongmen** [comment]
> > exciting shit! i have a mac now, though it might be a bit older than is desirable. tell me when you think i ought start looking at this, and i can try to build it. it'll likely take me some time to get there, as i've never developed on a mac.
> 
> Is it the mac mentioned in [#1280 (comment)](https://github.com/dankamongmen/notcurses/pull/1280#issuecomment-780911308)?

aye. it's a macbook pro, but a slightly older one

**Comment by michaelsbradleyjr** [comment]
> > Is it the mac mentioned in [#1280 (comment)](https://github.com/dankamongmen/notcurses/pull/1280#issuecomment-780911308)?
> 
> aye. it's a macbook pro, but a slightly older one

I'll share some information I've learned, some of it fairly recently despite using macOS (and Windows and Linux) over many years. It's not good news re: your mac hardware and I'm not recommending a particular course of action, just hoping it gives you more context.

For years now, Apple doesn't officially EOL versions of macOS (previously OS X). Instead, various software vendors and open source projects keep an eye on security patches that Apple releases over time. If Apple was previously releasing security patches for `10.N`, `10.N+1`, and `10.N+2`, but then in the next round `10.N` doesn't get patches, then "everybody knows" that means `10.N` is effectively EOL'd. The schedule isn't completely haphazard, their OS releases tend to be supported for about 3 to 4 years. `10.13` went EOL in January this year.

Homebrew is by far the most popular package manager for macOS, and it's a nice tool, though it has some limitations. As its website states:
> Homebrew is provided free of charge and run entirely by volunteers in their spare time.

Over time, its maintainers have pruned features and reduced functionalities you might expect in a package manager, to help keep `brew` simpler and easier to maintain. Also, they only maintain packages (termed "bottles") for non-EOL'd macOS versions. So, at present, any macOS version older than 10.14 won't be able to install packages with `brew`. The situation wasn't quite so severe until recently, because bottles for older macOS weren't being deleted, but then bintray shutdown this Spring and when Homebrew moved all the bottles to their new home (GitHub Packages), they didn't bring along any for macOS `< 10.14`.

Long-story short, your old mac hardware can't run `>= 10.14` so that's a no-go.

---

Ideas for ways forward:

* Those of us with more recent mac hardware and OS can contribute as best we can. It's harder for those of us (e.g. myself) who, despite interest in the project, are novices re: some of the concepts and C programming.
* GitHub Actions allows you to run up-to-date macOS environments, and it's not too difficult to setup (see e.g. [`status-im/nim-sqlcipher/blob/master/.github/workflows/test.yml`](https://github.com/status-im/nim-sqlcipher/blob/master/.github/workflows/test.yml#L24-L30). Unfortunately, it's not exactly an interactive environment, and so not very efficient for developing on a platform you're not familiar with.
* If there's enough community interest but not enough community know-how, maybe sponsorship could be scrapped together for purchase of a newer Mac mini (or something like that) for you to work with.
* I'm not sure how realistic it would be, but maybe some pair/group programming sessions, with you remotely driving a willing developer's mac. Could be fun, could be weird, dunno.

One other thing to note. Apple's hardware lineup now consists of laptops, "minis", etc. that feature their M1 ARM architecture. They're also still selling Intel based hardware. Things are getting sorted out slowly re: compilers, etc., but there are definitely some pain points that a lot of us are hitting during the transition period.

**Comment by dankamongmen** [comment]
> Long-story short, your old mac hardware can't run `>= 10.14` so that's a no-go.

ten thousand curses! this machine is of less than no use to me, then. i shall destroy it with thermite and AR-15 rounds.

thanks for the walkthrough and great information.


so the github actions solution seems pretty great. i'm not sure what the cost would be there, but i'd be willing to unass a few bucks a month. ideally it would run in CI on every commit, as the other CI does now, but if necessary, it could just gate releases.

wrt a mac mini, i'm...financially very lucky lol, and can just buy one. my hesitancy to do so stems from 25 years of agitating against Apple and its disneyfication of computing. like, what's this shit about my older laptop not being able to build things, wtf.

i'm happy to answer any emails or questions (ideally posted to the discussions area), and try to stay in the matrix channel. i'm very interested in mac support, and would be inclined to provide as much of my time as is necessary (subject to work and general baller constraints), especially if it's done in a way that ought keep it working going forward.

so in summary, let me kinda reverse your suggestion: rather than the community funding mac access for me to flail around on, i'd like to personally fund (if necessary) github-based macos CI that someone else [poke] sets up. i am 40, and just started working at MSFT and already have to deal with that ignominy and annoyance, and my desire to also learn MacOSX is, like, i'd rather learn OSX than be bitten on the balls repeatedly by moray eels, but i'd probably rather be pelted with eggs than either.

**Comment by dankamongmen** [comment]
btw, you're from STL? i was there twice in 2019 and thought it a **bangin** city, would definitely consider living there.

**Comment by michaelsbradleyjr** [comment]
> > Long-story short, your old mac hardware can't run `>= 10.14` so that's a no-go.
> 
> ten thousand curses! this machine is of less than no use to me, then. i shall destroy it with thermite and AR-15 rounds.

Sounds fun. 🧨 
 
> so the github actions solution seems pretty great. i'm not sure what the cost would be there, but i'd be willing to unass a few bucks a month. ideally it would run in CI on every commit, as the other CI does now, but if necessary, it could just gate releases.

GitHub Actions (GHA) is free, basically, for open source projects, unless you run up the usage super high, but given how relatively quickly notcurses compiles, I don't think you'll ever come near hitting the limit (the limit is documented somewhere just can't find it atm).

With a [config like this](https://github.com/status-im/nim-sqlcipher/blob/master/.github/workflows/test.yml#L3-L7) it would run on every commit made to an open PR (draft or normal) and also on every commit to the `master` branch. Other configurations are possible, of course.

> so in summary, let me kinda reverse your suggestion: rather than the community funding mac access for me to flail around on, i'd like to personally fund (if necessary) github-based macos CI that someone else [poke] sets up. i am 40, and just started working at MSFT and already have to deal with that ignominy and annoyance, and my desire to also learn MacOSX is, like, i'd rather learn OSX than be bitten on the balls repeatedly by moray eels, but i'd probably rather be pelted with eggs than either.

I'd be happy to submit a PR with a GHA workflow, shouldn't take long. What would you want it to do at first, just attempt compilation? (I'm not sure if `notcurses-tester` will run correctly in CI, but maybe it would).

Do you want it limited to Linux and macOS? Or do you want me to throw Windows + [MSYS2](https://www.msys2.org/) in the mix as well?

I played around with the latter yesterday, though no joy — I think it's going to require more conditionals in the headers, etc. before notcurses will compile in an MSYS2 context. Note that after compiling in an MSYS2 bash shell, if `-static` was passed to the compiler so the executable is truly standalone, the executable should be runnable from PowerShell or cmd.exe, though needs to be run in a terminal emulator like [ConEmu](https://conemu.github.io/) or the newish Windows Terminal. The reason I know this is that I've had luck building a `-static` exe with MSYS2 ncurses on Windows 10, copied it to a bare-bones Windows 7 x64 VM, and it ran perfectly in ConEmu.

> i am 40

We're of the same era, I'm 44. 🙂

**Comment by michaelsbradleyjr** [comment]
> btw, you're from STL? i was there twice in 2019 and thought it a **bangin** city, would definitely consider living there.

I'm a transplant, moved to STL in 2008, but it's definitely home now, I intend to stay here for the long haul. It's a "big little town", for sure — not a huge population in the city proper anymore, but the larger metro era is big in size and population. We've got a symphony, museums, largest city park in the country (even bigger than NYC's), world-class zoo, Cardinals and Blues, historic architecture (it's an old city, e.g. my house was built in 1870 and that's not ucommon), multiple universities, all sorts of stuff. We've got problems too: crime is the worst since I moved here, and I choose to live in the city, so... well I could tell you some stories; even so, my friends (some natives, some transplants like me) love it and look forward to brighter years ahead. 🤞

**Comment by o-sdn-o** [comment]
@dankamongmen Have you considered using macOS on VirtualBox? Performance with this approach is low, but it is quite enough for testing/compiling code under macOS. https://github.com/myspaghetti/macos-virtualbox

**Comment by dankamongmen** [comment]
is this about ready to go? i'm gonna go ahead and review if so!

**Comment by dankamongmen** [comment]
by the way, if it isn't clear, there are parts that don't need perfect macos support to get merged. in particular, i'm thinking of the subprocess stuff; we use pidfds there on linux, which definitely don't exist on mac (afaik). so don't worry about deep system stuff like that. i or someone else can handle it once merged. i'd love to see this land soon.

**Comment by michaelsbradleyjr** [comment]
> by the way, if it isn't clear, there are parts that don't need perfect macos support to get merged. in particular, i'm thinking of the subprocess stuff; we use pidfds there on linux, which definitely don't exist on mac (afaik). so don't worry about deep system stuff like that. i or someone else can handle it once merged. i'd love to see this land soon.

Beyond working on the initial GitHub Actions workflow and then reviewing @igo95862's revision in #1764, I haven't had time to circle back around to actually work on the macOS compatibility aspects of notcurses. I'll rebase this branch against current `master` branch and see how it goes locally and in GHA. If I recall correctly, from rebasing a few days ago, the macOS build now doesn't get past `make`, whereas it used to build completely but fail during `ctest`.

Maybe we can loop @grendello back in, if he still has interest. Also, in fairly recent comments in #1280, @kdheepak expressed interest and I encouraged him to take a look at this PR's branch instead of #1280's (because this one is more up-to-date), but I'm not sure if he had an opportunity or any luck.

**Comment by dankamongmen** [comment]
> > by the way, if it isn't clear, there are parts that don't need perfect macos support to get merged. in particular, i'm thinking of the subprocess stuff; we use pidfds there on linux, which definitely don't exist on mac (afaik). so don't worry about deep system stuff like that. i or someone else can handle it once merged. i'd love to see this land soon.
> 
> Beyond working on the initial GitHub Actions workflow and then reviewing @igo95862's revision in #1764, I haven't had time to circle back around to actually work on the macOS compatibility aspects of notcurses. I'll rebase this branch against current `master` branch and see how it goes locally and in GHA. If I recall correctly, from rebasing a few days ago, the macOS build now doesn't get past `make`, whereas it used to build completely but fail during `ctest`.
> 
> Maybe we can loop @grendello back in, if he still has interest. Also, in fairly recent comments in #1280, @kdheepak expressed interest and I encouraged him to take a look at this PR's branch instead of #1280's (because this one is more up-to-date), but I'm not sure if he had an opportunity or any luck.

is there any way i can get a shell on your macos environment?

**Comment by dankamongmen** [comment]
> > Maybe we can loop @grendello back in, if he still has interest. Also, in fairly recent comments in #1280, @kdheepak expressed interest and I encouraged him to take a look at this PR's branch instead of #1280's (because this one is more up-to-date), but I'm not sure if he had an opportunity or any luck.
> 
> is there any way i can get a shell on your macos environment?

if it's just laptops, it's probably not worth it, but if you have even a wimpy always-up mac machine, i'm happy to get in there this weekend and see what i can do. https://nick-black.com/authorized_keys

**Comment by michaelsbradleyjr** [comment]
My personal one (i.e. on my laptop), or in the GitHub Actions environment?

If you meant the latter, the answer is no, just doesn't work that way as far as I know.

If the former... ah, you replied in the meantime...

It's just a laptop, pretty personal device, and I also have stuff on it related to my employment, etc.

I'll look into bringing a "clean slate" Mac mini online, but it's not something I can make happen this weekend, since I don't actually have it at my house currently.

**Comment by michaelsbradleyjr** [comment]
> I'll look into bringing a "clean slate" Mac mini online, but it's not something I can make happen this weekend, since I don't actually have it at my house currently.

I should have it at my house by July 5–9.  I'll blank it out, install up-to-date macOS and Xcode, use [`brew`](https://brew.sh/) to install various build tools and whatnot, create the `dank@` user and enable ssh access per the key you linked.

Hopefully it's still helpful by then.

**Comment by michaelsbradleyjr** [comment]
With a little hackery I got `make` to complete successfully again on macOS, up-to-date with latest changes on `master` (as of when I started on it an hour+ ago).

Changes required: https://github.com/michaelsbradleyjr/notcurses/commit/b9e8a47dde41ded335c277782188523a026b82c0

Some of it's obvious, I think; I'll provide more context for things that might not be:

I had to add `-lreadline` to `DCMAKE_C_FLAGS` when invoking `cmake`. I don't know why it's required now and not before, but the linker couldn't succeed otherwise.

In `linux.c` I updated the definition for `is_linux_console` but had to comment out `(void)nc;`, otherwise the compiler complained `nc` wasn't defined.

In `interp.c` I added some `include`s that were suggested in a web search result; apparently, that's the required incantation on macOS if you intend to `#include <sys/random.h>`.

Also re: `interp.c`, `getrandom` doesn't seem to be part of `sys/random.h` on macOS; I saw [your comment re: the qr code demo](https://github.com/dankamongmen/notcurses/blob/master/src/demo/qrcode.c#L18), so tried doing something similar; at least it allowed the compiler to succeed.

See this macOS GHA run per my fork:

https://github.com/michaelsbradleyjr/notcurses/runs/2923234329?check_suite_focus=true

It fails in `ctest`, as before. Locally, I can run the demo without crash so long as I leave `arw` out of the demospec.
* `a` causes the process to exit with error
* `r` runs, looks okay (relative to e.g. ubuntu+kitty), but eventually hangs: cpu is at 100% and the demo is frozen, have to ctrl+c
* `w` freezes immediately, have to ctrl+c

**Comment by michaelsbradleyjr** [comment]
Happily, after rebasing again re: `master` , all the changes in [michaelsbradleyjr@`b9e8a47`](https://github.com/michaelsbradleyjr/notcurses/commit/b9e8a47dde41ded335c277782188523a026b82c0) became unnecessary, so they're now gone in my [`michaelsbradleyjr:macos`](https://github.com/michaelsbradleyjr/notcurses/tree/macos) branch.

Relative to https://github.com/dankamongmen/notcurses/pull/1701#issuecomment-869085700, almost everything can be considered irrelevant, but here's the update for things still relevant:

> It fails in ctest, as before. Locally, I can run the demo without crash so long as I leave arw out of the demospec.
> 
> a causes the process to exit with error
> ~~r runs, looks okay (relative to e.g. ubuntu+kitty), but eventually hangs: cpu is at 100% and the demo is frozen, have to ctrl+c~~
> w freezes immediately, have to ctrl+c

Latest GHA run for macOS: https://github.com/michaelsbradleyjr/notcurses/runs/2923362196?check_suite_focus=true.

Interestingly, the ubuntu run also fails: https://github.com/dankamongmen/notcurses/runs/2923361254?check_suite_focus=true.

**Comment by michaelsbradleyjr** [comment]
> I should have it at my house by July 5–9. I'll blank it out, install up-to-date macOS and Xcode, use [`brew`](https://brew.sh/) to install various build tools and whatnot, create the `dank` user and enable ssh access per the key you linked.
> 
> Hopefully it's still helpful by then.

The Mac mini is arriving a bit earlier than expected, should be here later today.

If I have enough energy after work, I'll set it up this evening; otherwise, I'll do it tomorrow.

**Comment by dankamongmen** [comment]
> The Mac mini is arriving a bit earlier than expected, should be here later today.
> If I have enough energy after work, I'll set it up this evening; otherwise, I'll do it tomorrow.

you are THE MAN. give me a paypal address or a venmo and i'll send some cash your way. AWESOIME.

**Comment by michaelsbradleyjr** [comment]
> > The Mac mini is arriving a bit earlier than expected, should be here later today.
> > If I have enough energy after work, I'll set it up this evening; otherwise, I'll do it tomorrow.
> 
> you are THE MAN. give me a paypal address or a venmo and i'll send some cash your way. AWESOIME.

**Comment by michaelsbradleyjr** [comment]
Whoop, clicked the wrong button.

Anyway, email sent with details re: the Mac mini.

**Comment by dankamongmen** [comment]
> Anyway, email sent with details re: the Mac mini.

got it! responding now =]

**Comment by dankamongmen** [comment]
i've got some time to work on this now, and am doing so. i'll take this and start merging in chunks, as i understand them. heroic work, @michaelsbradleyjr !

**Comment by michaelsbradleyjr** [comment]
I'd say the bulk of the programming work was done by @grendello, I just massaged it until it started working again. :)

File: src/lib/macos-compat.c [hunk 1]
```diff
@@ -0,0 +1,100 @@
+#include <fcntl.h>
+#include <pthread.h>
+#include <signal.h>
+#include <unistd.h>
+#include <time.h>
+
+#include "macos-compat.h"
+
+// Time code based on: https://github.com/ChisholmKyle/PosixMachTiming/tree/master/src
```
Comment (dankamongmen) [inline src/lib/macos-compat.c:9]: this is ISC-licensed, so bringing it into our tree would cause some licensing-related complications. i'll put together anything we need of this type. thanks for proving it out, though!

File: CMakeLists.txt [hunk 2]
```diff
@@ -76,6 +82,51 @@ add_compile_options(-Wall -Wextra -W -Wshadow -Wformat -Wno-nonnull-compare
...
+
+  # ncurses version shipped with the latest macOS SDK is a fossilware version of 5.7 (A.D. 2008)
+  # We need 6.1, we need to take it from brew (for instance) and we need to be
+  # sneaky since the newer versions must not disrupt the fossilware installed with
+  # the SDK
+
+  # We support homebrew (https://brew.sh/) for now since I don't use anything else
+  find_program(BREW brew REQUIRED)
+
+  # We set up the environment so that pkg-config below can find ncurses
```
Comment (dankamongmen) [inline CMakeLists.txt:104]: this seems kinda heavyweight for what we need. is it not sufficient to add /usr/local/include and /usr/local/lib to the include/link paths in CMake, and then use the same routines we do for everything else? i really hate the idea of updating things in two places, especially in shitty shitty CMake

Comment (dankamongmen) [reply CMakeLists.txt:104]: i'll see what can be done, but if you have particular reasoning why you went this direction, i'd like to hear it

File: CMakeLists.txt [hunk 3]
```diff
@@ -140,7 +191,7 @@ file(GLOB COMPATSRC CONFIGURE_DEPENDS src/compat/compat.c)
 ############################################################################
 # libnotcurses-core (core shared library, core static library)
 file(GLOB NCCORESRCS CONFIGURE_DEPENDS src/lib/*.c src/lib/*.cpp)
-add_library(notcurses-core SHARED ${NCCORESRCS} ${COMPATSRC})
+add_library(notcurses-core SHARED ${NCCORESRCS} ${COMPATSRC} ${MACOS_COMPAT_SRC})
```
Comment (dankamongmen) [inline CMakeLists.txt:194]: i'd rather just use the existing compat.c and #ifdef within there, rather than need this everywhere.

File: CMakeLists.txt [hunk 4]
```diff
@@ -611,14 +663,23 @@ target_include_directories(notcurses-tester
...
   PRIVATE
     notcurses++
     "${unistring}"
     "${TERMINFO_LIBRARIES}"
+    doctest::doctest
 )
+if(APPLE)
+  target_link_directories(notcurses-tester
+    PRIVATE
+      "${TERMINFO_LIBRARY_DIRS}"
```
Comment (dankamongmen) [inline CMakeLists.txt:679]: hrmm. i'd expect these to fall out from the pkg-config discoveries earlier.

**Review by dankamongmen — COMMENTED** [review]
no need to make these changes, just asking a few questions

**Comment by dankamongmen** [comment]
yay, i have `ncneofetch` working on OSX
```
                ───────────────────────────[ dank@cobra-la ]────────────────────────────
                │Darwin 20.5.0                                                         │
                │                                                                      │
                │DM: (null)                                  Shell: /usr/local/bin/bash│
                │RGB TERM: Kitty 0.21.2                                    Screen0: 0x0│
                │LANG: en_US.UTF-8                                             UID: 502│
                │                        Macmini8,1 (12 cores)                         │
                ╰──────────────────────────────────────────────────────────────────────╯
```

**Comment by dankamongmen** [comment]
everything now builds on OSX in the `dankamongmen/osx-safety-brain` branch, with the following:

`PKG_CONFIG_PATH="/usr/local/Cellar/ncurses/6.2/lib/pkgconfig:/usr/local//Cellar/readline/8.1/lib/pkgconfig"`

**Comment by michaelsbradleyjr** [comment]
Should this one be closed in favor of #1923?

File: CMakeLists.txt (hunk 2 above)
Comment (michaelsbradleyjr) [reply CMakeLists.txt:104]: macOS ships with old (sometimes ancient) versions of software such as ncurses, bash, etc.

In order not to conflict with the stuff that ships with the OS, `brew` organizes things in `/usr/local/opt` (which is a collection of symlinks to specific versions in `/usr/local/Cellar`), and you get these kind of messages with `brew info`:

```
$ brew info ncurses
...

ncurses is keg-only, which means it was not symlinked into /usr/local,
because macOS already provides this software and installing another version in
parallel can cause all kinds of trouble.

If you need to have ncurses first in your PATH, run:
  echo 'export PATH="/usr/local/opt/ncurses/bin:$PATH"' >> /Users/michael/.bash_profile

For compilers to find ncurses you may need to set:
  export LDFLAGS="-L/usr/local/opt/ncurses/lib"
  export CPPFLAGS="-I/usr/local/opt/ncurses/include"

For pkg-config to find ncurses you may need to set:
  export PKG_CONFIG_PATH="/usr/local/opt/ncurses/lib/pkgconfig"

...
```

While it's not the only possibility, pretty much everyone doing open source on macOS consumes Homebrew packages and/or publishes such packages.

Comment (michaelsbradleyjr) [reply CMakeLists.txt:104]: Sorry if that's a repeat from something I explained in my emails; main point is that on macOS it's more likely to work more easily for more devs and users if its done the Homebrew way.

**Comment by michaelsbradleyjr** [comment]
Closing in light of #1923. (if it shouldn't be closed yet then it can be reopened).
